### ISSUE TITLE
Store Query error: `get range reader: The specified key does not exist`

### ISSUE THREAD

**Issue by haeringer** [thanos-io/thanos#4907]
**Thanos, Prometheus and Golang version used**: Prometheus v2.30.3 and Thanos v0.23.1 on Kubernetes

**Object Storage Provider**: S3

**What happened**: After having shipped all existing metrics from a Prometheus instance to S3 via `--shipper.upload-compacted`, some of the uploaded metrics are not retrievable because queries return the following error:
```
Error executing query: expanding series: proxy Series(): Addr: thanos-store-prod:10901 LabelSets: {cluster="kubernetes_cluster_name", datacenter="datacenter_name", prometheus="prometheus_instance_name"},{datacenter="eu-central-1a", prometheus="example"} Mint: 1629460800027 Maxt: 1637913600000: receive series from Addr: thanos-store-prod:10901 LabelSets: {cluster="kubernetes_cluster_name", datacenter="datacenter_name", prometheus="prometheus_instance_name"},{datacenter="eu-central-1a", prometheus="example"} Mint: 1629460800027 Maxt: 1637913600000: rpc error: code = Aborted desc = fetch series for block 01FMED6TH7XWQK5SY2DZWQSS3P: load chunks: get range reader: The specified key does not exist.
```

With "some metrics" I mean lots of the uploaded blocks which are somewhere in between other compacted uploaded blocks that are queryable without error.

**What you expected to happen**: Receiving the metrics without an error

**How to reproduce it (as minimally and precisely as possible)**: I don't know, because it did happen on this particular set of data.

**Full logs to relevant components**: Haven't found any logs related to this error. Maybe the meta.json of an affected block is of interest?:

<details>meta.json
<p>

```
{
  "ulid": "01FMED6TH7XWQK5SY2DZWQSS3P",
  "minTime": 1636653600047,
  "maxTime": 1636848000000,
  "stats": {
    "numSamples": 2851641134,
    "numSeries": 546111,
    "numChunks": 20023942
  },
  "compaction": {
    "level": 4,
    "sources": [
      "01FM8CSPDGHB9PKAW4C1E8KKDE",
      "01FM8KNDNG3EJW06BSWGJ98CZM",
      "01FM8TH4XHSY6Z4THZVX36WMF5",
      "01FM91CW5GX4XKFBBH7YDGJ8S7",
      "01FM988KJ9NGBVXRRMBMTJKJ5H",
      "01FM9F4ANHNR70G9S5QETJS9ZT",
      "01FM9P01XHZW1C5H9MK4ZEJ7BY",
      "01FM9WVS5HBJ1SGNP1JM6ZH0AZ",
      "01FMA3QGDGHWMTWZHJT9180Q33",
      "01FMAAK7NG6C2AR21EH1MPR4BN",
      "01FMAHEYXJX7VM5W6NK91A36GS",
      "01FMARAP5GFCTGH5CQHE36YD65",
      "01FMAZ6DDHF0T12WDQ2281918D",
      "01FMB624NGJ63H9KY8CZJ67PGC",
      "01FMBCXW28FJNE5F7EW2BN4212",
      "01FMBKSKGDSX7GAT93XCEDYCV6",
      "01FMBTNADGQD95KAEG5BTKKQHY",
      "01FMC1H1NGNMD0WEVM0BV0TBZH",
      "01FMC8CRXGZBEPTN5XXFAF5JHS",
      "01FMCF8G5GK1RDBA6E4C6ENCE7",
      "01FMCP47DHYD8VT1JBCBGADYP7",
      "01FMCWZYNHD91A2S60ZZV15DGP",
      "01FMD3VNXGGKCKS355WH3NQ3AM",
      "01FMDAQD5F8GTJJPWZ87YWKAG8",
      "01FMDHK4DG6XPWXQAG91KPJ3WW",
      "01FMDREVNGX3EN2F0861RHTDHW",
      "01FMDZAJXGNYF2DYDM3PWQVB8T"
    ],
    "parents": [
      {
        "ulid": "01FMAHG9NBYT11KBEKPTMJ8V7G",
        "minTime": 1636653600047,
        "maxTime": 1636718400000
      },
      {
        "ulid": "01FMCFB6YA44D1DRMR3ZKRNPE6",
        "minTime": 1636718400195,
        "maxTime": 1636783200000
      },
      {
        "ulid": "01FMED4TE39QFRBW2S9KAQ7SWZ",
        "minTime": 1636783200105,
        "maxTime": 1636848000000
      }
    ],
    "failed": true
  },
  "version": 1,
  "thanos": {
    "labels": {
      "datacenter": "eu-central-1a",
      "prometheus": "example"
    },
    "downsample": {
      "resolution": 0
    },
    "source": "sidecar",
    "segment_files": [
      "000001",
      "000002",
      "000003",
      "000004",
      "000005",
      "000006",
      "000007",
      "000008"
    ],
    "files": [
      {
        "rel_path": "chunks/000001",
        "size_bytes": 536869880
      },
      {
        "rel_path": "chunks/000002",
        "size_bytes": 536870555
      },
      {
        "rel_path": "chunks/000003",
        "size_bytes": 536870679
      },
      {
        "rel_path": "chunks/000004",
        "size_bytes": 536870766
      },
      {
        "rel_path": "chunks/000005",
        "size_bytes": 536870705
      },
      {
        "rel_path": "chunks/000006",
        "size_bytes": 536870546
      },
      {
        "rel_path": "chunks/000007",
        "size_bytes": 536870812
      },
      {
        "rel_path": "chunks/000008",
        "size_bytes": 162057689
      },
      {
        "rel_path": "index",
        "size_bytes": 245227371
      },
      {
        "rel_path": "meta.json"
      }
    ]
  }
}
```

</p>
</details>

Please let me know if there is any additional information I can gather.

**Comment by apoorva-marisomaradhya** [comment]
@haeringer Were you able to resolve this issue?

**Comment by haeringer** [comment]
No, since this happened after we migrated old, locally stored metrics to S3, the issue eventually became obsolete and we decided not to pursue it further.

**Comment by stBlackD** [comment]
After updating to 0.30.1, we also encountered this problem.
@apoorva-marisomaradhya did you manage to solve it?

**Comment by levya4** [comment]
I encountered the exact same issue :(

the error after querying the old data:
`fetch series for block 01GSJ3EQQ1N7G2DR1WPFK8W27G: load chunks: get range reader: The specified key does not exist." msg="returning partial response"`

Did someone was able to solve it?

**Comment by dimm0** [comment]
Same here

**Comment by jbkc85** [comment]
Same here.  Curious on the fix for this particular issue?

**Comment by tizki** [comment]
anyone was able to resolve this? I have the same issue.

**Comment by marcolenzo** [comment]
Same issue here but I can add some detail to it.

In my scenario, I noticed that the particular block triggering the error (`01J5NJDXD8J3VY3QJ4QDMADW9S`) was present on object storage but it was missing the `chunks` folder. On the other hand, the chunks existed on the volume attached to prometheus. This made me think something went wrong during the upload.

**The way to fix it is to retrigger the problematic block upload**. I don't know if I used the correct procedure but what I did was:

- delete the folder from object storage
- modify the `thanos.shipper.json` (which exists in the prometheus folder) by removing the problematic block from the `uploaded` entries.
- restarted the prometheus pod

```
ts=2024-08-28T07:44:06.307616573Z caller=shipper.go:372 level=info msg="upload new block" id=01J5NJDXD8J3VY3QJ4QDMADW9S
```

That solved it for me! Hope this helps.

That said, I have no idea why the chunks were missing in the first place.
